content/post/2019-01-20-what-is-production-grade.markdown

```diff
@@ -0,0 +1,270 @@
+unhappiness about the system state does not demand action.
+
+Put more simply and more directly: if it is important enough to wake
+some one up and put a keyboard in front of them, its part of
+production.
+
+With a reasonable definition of what production is, lets now look at
+what production grade means.
+
+## What is Production Grade?
+
+Like lumber, meat, and equipment, its not unreasonable to say that
+software comes in various grades.  In many cases these terms are
+legally protected.  For example the United States Department of
+Agriculture governs the use of specific terms for grades of meat.
+Similarly the Western Wood Products Association provides grading
+criteria for lumber.  It is interesting then that no such set of
+standards exists for software.  Anyone can claim at any time that any
+software is production grade, or is of sufficient quality to wake
+someone up at night.
+
+Its at this point that I'm going to clearly state that for the
+purposes of this article, I'm glossing over the complexity of software
+fitness for a particular purpose, such as life or safety critical
+systems.  These systems usually require the review and certification
+of a Professional Engineer, whose title and license are part of a
+regulated profession in most parts of the world.  I am not a
+Professional Engineer, and so I can't speak to the kinds of things
+that one would look for in certifying software.  I can only speak to
+what I look for in things that I allow to wake me up at night.
+
+Like most things, this is a sliding scale.  Lets start at the huge end
+and work our way down.
+
+### Insane Scale
+
+During my tenure at Google, I had a very specific definition of
+something being in production since I was an SRE.  This definition
+boiled down to "can it page me?"  Since I was part of a Tier 1
+rotation, by definition we would only be on call for things that were
+in production, and by definition they should have been production
+grade (whether or not this was true is a debate for another time).
+
+In this environment, I personally held the definition that production
+grade wasn't something I could write down.  It was something I knew
+when I saw it, and was greater than the sum of its parts.  Those
+familiar with US Supreme Court decisions might find this equivalently
+infuriating to Justice Potter Stewart's test for obscenity.  When a
+developer wanted me to OK a feature to push to the many many tasks we
+had running around the world I had a checklist for certain, but I also
+had to ask myself what my gutt feeling was on the change.  Was it
+safe?  Would it cause downtime?  How might this affect disaster
+response if we needed to shut off the new widget?  All of these are
+important questions to ask when looking for "Production Grade" at
+insane scales.
+
+Some of the biggest things I looked for had to do with startup and
+shutdown.  I really wanted to see that in anything I was expected to
```

Comment:
I'd move `in anything I was expected to be responsible for` to the start of the sentence.